File: Data Structures & Algorithms/combination-target-sum/submission-21.js
class Solution {
    /**
     * @param {number[]} nums
     * @param {number} target
     * @returns {number[][]}
     */
    combinationSum(nums, target) {  
        const res = [] 
        nums.sort() 

        function dfs(start, total, cur) {
            if (total === target) {
                res.push([...cur]) 
                return
            } 

            for (let i = start; i < nums.length; i ++) { 
                if (total + nums[i] > target) return
                cur.push(nums[i]) 
                dfs(i, total + nums[i], cur) 
                cur.pop()
            }
        } 

        dfs(0, 0, []) 

        return res
    }
}

File: Data Structures & Algorithms/insert-new-interval/submission-21.py
class Solution:
    def insert(self, intervals: List[List[int]], newInterval: List[int]) -> List[List[int]]:
        res = []  
        i = 0  

        start, end = newInterval

        while i < len(intervals): 
            cur = intervals[i] 

            if cur[1] < start: 
                res.append(cur) 
                i += 1 
                continue 

            if end >= cur[0]: 
                start = min(start, cur[0]) 
                end = max(end, cur[1])  
                i += 1 
                continue  

            break

        res.append([start, end]) 

        while i < len(intervals): 
            res.append(intervals[i]) 
            i += 1 

        return res

